lib/interceptor/cache.js

```diff
@@ -0,0 +1,155 @@
+'use strict'
+
+const CacheHandler = require('../handler/cache-handler')
+const MemoryCacheStore = require('../cache/memory-cache-store')
+const CacheRevalidationHandler = require('../handler/cache-revalidation-handler')
+const { UNSAFE_METHODS } = require('../util/cache.js')
+
+const AGE_HEADER = Buffer.from('age')
+
+/**
+ * @param {import('../../types/cache-interceptor.d.ts').default.CacheOptions | undefined} globalOpts
+ * @returns {import('../../types/dispatcher.d.ts').default.DispatcherComposeInterceptor}
+ */
+module.exports = globalOpts => {
+  if (!globalOpts) {
+    globalOpts = {}
+  }
+
+  if (globalOpts.store) {
+    for (const fn of ['get', 'put']) {
+      if (typeof globalOpts.store[fn] !== 'function') {
+        throw new TypeError(`CacheStore needs a \`${fn}()\` function`)
+      }
+    }
+
+    for (const getter of ['entryCount', 'maxEntries', 'maxEntrySize']) {
+      const actualType = typeof globalOpts.store[getter]
+      if (actualType !== 'number') {
+        throw new TypeError(`CacheStore needs a ${getter} property with type number, current type: ${actualType}`)
+      }
+    }
+
+    for (const value of ['maxEntries', 'maxEntry']) {
+      if (globalOpts.store[value] <= 0) {
+        throw new TypeError(`CacheStore ${value} needs to be >= 1`)
+      }
+    }
+  } else {
+    globalOpts.store = new MemoryCacheStore()
+  }
+
+  if (globalOpts.methods) {
+    if (!Array.isArray(globalOpts.methods)) {
+      throw new TypeError(`methods needs to be an array, got ${typeof globalOpts.methods}`)
+    }
+
+    if (globalOpts.methods.length === 0) {
+      throw new Error('methods must have at least one method in it')
+    }
+  } else {
+    globalOpts.methods = ['GET']
+  }
+
+  // Safe methods the user wants and unsafe methods
+  const methods = [...globalOpts.methods, ...UNSAFE_METHODS]
+
+  return dispatch => {
+    return (opts, handler) => {
+      if (!opts.origin || !methods.includes(opts.method)) {
+        // Not a method we want to cache or we don't have the origin, skip
+        return dispatch(opts, handler)
+      }
+
+      const stream = globalOpts.store.createReadStream(opts)
+      if (!stream) {
+        // Request isn't cached
+        return dispatch(opts, handler)
+      }
+
+      const { value } = stream
+
+      // Dump body on error
+      if (typeof opts.body === 'object' && opts.body.constructor.name === 'Readable') {
+        opts.body?.on('error', () => {}).resume()
+      }
```

Comment:
You can use the internal `Redeable`. It already comes with `dump` up to a watermark API

Reply:
Wdym?

Reply:
IMHO this is fine